tests/test_queryselector.py

```diff
@@ -0,0 +1,309 @@
+
+    @chrome_only
+    @sync
+    async def test_Jeval_throws_exc_if_no_element(self, isolated_page):
+        """Test Page().Jeval() should throw error if no element is found."""
+        page = isolated_page
+        with pytest.raises(ElementHandleError, match='Error: failed to find element matching selector "section"'):
+            await page.Jeval('section', "e => e.id")
+
+
+class TestPageJJeval:
+
+    @chrome_only
+    @sync
+    async def test_JJeval_executes_js_func(self, isolated_page):
+        """Test Page().JJeval() method executes JS function
+        with all elements which matches ``selector``.
+
+        """
+        page = isolated_page
+        await page.setContent('<div>hello</div><div>beautiful</div><div>world!</div>')
+        divsCountJJeval = await page.JJeval('div', "divs => divs.length")
+        divsCountQuery = await page.querySelectorAllEval('div', "divs => divs.length")
+        assert divsCountJJeval == divsCountQuery == 3
+
+
+class TestPageJ:
+
+    @chrome_only
+    @sync
+    async def test_J_queries_element(self, isolated_page):
+        """Test Page().J() method should query existing element."""
+        page = isolated_page
+        await page.setContent('<section>test</section>')
+        element = await page.J('section')
+        assert element
+
+    @chrome_only
+    @sync
+    async def test_J_returns_none_if_no_element(self, isolated_page):
+        """Test Page().J() method should return null for non-existing element."""
+        page = isolated_page
+        element = await page.J('non-existing-element')
+        assert element is None
+
+
+class TestPageJJ:
+
+    @chrome_only
+    @sync
+    async def test_JJ_queries_elements(self, isolated_page):
+        """Test Page().JJ() method should query existing elements."""
+        page = isolated_page
+        await page.setContent('<div>A</div><br/><div>B</div>')
+        elements = await page.JJ('div')
+        assert len(elements) == 2
+        assert [await page.evaluate("e => e.textContent", element) for element in elements] == ['A', 'B']
+
+    @chrome_only
+    @sync
+    async def test_JJ_returns_empty_list_if_no_element(self, isolated_page, server):
+        """Test Page().JJ() method should return empty list if nothing is found."""
+        page = isolated_page
+        await page.goto(server / "empty.html")
+        elements = await page.JJ('div')
+        assert elements == []
+
+
+class TestPageJx:
+
+    @chrome_only
+    @sync
+    async def test_Jx_queries_elements(self, isolated_page):
+        """Test Page().Jx() method should query existing elements."""
+        page = isolated_page
+        await page.setContent('<section>test</section>')
+        elementsJx = await page.Jx('/html/body/section')
+        elementsXpath = await page.xpath('/html/body/section')
+        assert isinstance(elementsJx, list)
+        assert isinstance(elementsXpath, list)
+        assert elementsJx[0]
+        assert elementsXpath[0]
```

Comment:
Lookin for advice how we can compare `elementsJx[0]` and `elementsXpath[0]` by value?

Reply:
We can't really. The best way I could think of is comparing the values of `RemoteObject`, but even then we'd need to filter out the IDs that would be different. Here's what each object's `_remoteObject` property looks like:
```py
{'type': 'object', 'subtype': 'node', 'className': 'HTMLElement', 'description': 'section', 'objectId': '{"injectedScriptId":1,"id":6}'}
```
```py
{'type': 'object', 'subtype': 'node', 'className': 'HTMLElement', 'description': 'section', 'objectId': '{"injectedScriptId":1,"id":3}'}
```
I think the current testing method is sufficient, there are other places where the effectiveness of these is compared.

Reply:
Actually scrap that. A much better way to test all of this would be to just compare that the functions are the same:
```py
In[5]: page.Jx == page.xpath
Out[5]: True
```